include/net/lwm2m.h

```diff
@@ -365,6 +366,99 @@ struct coap_block_context *lwm2m_firmware_get_block_context();
 #endif
 #endif
 
+
+#if defined(CONFIG_LWM2M_SWMGMT_OBJ_SUPPORT)
+
+/**
+ * @brief Set callback to handle software activation requests
+ *
+ * This callback will get called when the Activate execute on
+ * the corresponding object instance is called and the object is
+ * in the correct state.
+ *
+ * @param[in] obj_inst_id The instance number to set the callback for.
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_activate_cb(uint16_t obj_inst_id, lwm2m_engine_user_cb_t cb);
+
+/**
+ * @brief Set callback to handle software deactivation requests
+ *
+ * This callback will get called when the Deactivate execute on
+ * the corresponding object instance is called and the object is
+ * in the correct state.
+ *
+ * @param[in] obj_inst_id The instance number to set the callback for.
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_deactivate_cb(uint16_t obj_inst_id, lwm2m_engine_user_cb_t cb);
+
+/**
+ * @brief Set callback to handle software install requests
+ *
+ * This callback will get called when the Install execute on
+ * the corresponding object instance is called and the object is
+ * in the correct state.
+ *
+ * @param[in] obj_inst_id The instance number to set the callback for.
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_install_package_cb(uint16_t obj_inst_id, lwm2m_engine_user_cb_t cb);
+
+/**
+ * @brief Set callback to handle software uninstall requests
+ *
+ * This callback will get called when the uninstall execute on
+ * the corresponding object instance is called and the object is
+ * in the correct state.
+ *
+ * @param[in] obj_inst_id The instance number to set the callback for.
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_delete_package_cb(uint16_t obj_inst_id, lwm2m_engine_user_cb_t cb);
+
+/**
+ * @brief Set callback to read software package
+ *
+ * This callback will get called during read on package version attribute on
+ * the corresponding object instance is called
+ *
+ * @param[in] obj_inst_id The instance number to set the callback for.
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_read_package_cb(uint16_t obj_inst_id, lwm2m_engine_get_data_cb_t cb);
```

Comment:
I just wonder, is this really needed? There's no special action needed on the object implementation side before the callback is exectued, so the app could simply use generic `lwm2m_engine_register_read_callback()`.

Reply:
Hmmm... on a functional level I don't think this is needed - meaning this doesn't provide unique functionality compared to `lwm2m_engine_register_read_callback`. The thing is that *read package* callback reads the resource's version attribute. It's not a typical *read the data callback*. So technically the user can implement reading the version attribute and pass that via the `lwm2m_engine_register_read_callback` but I think that would be a bit misleading. Having a separate function creates a distinction between *set read data callback* and *set read package version attribute*.

I hope I'm getting this right ;)